pilot/pkg/kube/inject/app_probe.go

```diff
@@ -77,45 +101,68 @@ func extractStatusPort(sidecar *corev1.Container) int {
 	return -1
 }
 
-// rewriteProbe changes application containers' probe to point to sidecar's status port.
-func rewriteProbe(probe *corev1.Probe, appProbers *status.KubeAppProbers,
-	newURL string, statusPort int, portMap map[string]int32) {
+// convertAppProber returns a overwritten `HTTPGetAction` for pilot agent to take over.
+func convertAppProber(probe *corev1.Probe, newURL string, statusPort int) *corev1.HTTPGetAction {
 	if probe == nil || probe.HTTPGet == nil {
-		return
+		return nil
 	}
-	httpGet := probe.HTTPGet
-
-	// Save app probe config to pass to pilot agent later.
-	savedProbe := proto.Clone(probe.HTTPGet).(*corev1.HTTPGetAction)
-	(*appProbers)[newURL] = savedProbe
-	// A named port, resolve by looking at port map.
-	if httpGet.Port.Type == intstr.String {
-		port, exists := portMap[httpGet.Port.StrVal]
-		if !exists {
-			log.Errorf("named port not found in the map skip rewriting probing %v", *probe)
-			return
-		}
-		savedProbe.Port = intstr.FromInt(int(port))
-	}
-	// Change the application csince ontainer prober config.
-	httpGet.Port = intstr.FromInt(statusPort)
-	httpGet.Path = newURL
+	c := proto.Clone(probe.HTTPGet).(*corev1.HTTPGetAction)
+	// Change the application container prober config.
+	c.Port = intstr.FromInt(statusPort)
+	c.Path = newURL
+	return c
 }
 
-func rewriteAppHTTPProbe(spec *SidecarInjectionSpec, podSpec *corev1.PodSpec) {
-	if spec == nil || podSpec == nil {
-		return
-	}
-	if !spec.RewriteAppHTTPProbe {
-		return
+// DumpAppProbers returns a json encoded string as `status.KubeAppProbers`.
+// Also update the probers so that all usages of named port will be resolved to integer.
+func DumpAppProbers(podspec *corev1.PodSpec) string {
+	out := status.KubeAppProbers{}
+	updateNamedPort := func(p *corev1.Probe, portMap map[string]int32) *corev1.HTTPGetAction {
+		if p == nil || p.HTTPGet == nil {
+			return nil
+		}
+		h := p.HTTPGet
+		if h.Port.Type == intstr.String {
+			port, exists := portMap[h.Port.StrVal]
+			if !exists {
+				return nil
```

Comment:
In this function, if a probe port likely 8080 is not specified in `podspec.Containers[].Ports`.  Then It will not rewrite the probe? If so, I think it's not right.

Reply:
This block is conditioned on `h.Port.Type == intstr.String` if your port is 8080, the type will be `intstr.Int`, regardless of whether that is declared or not.

We actually have a unit test to cover that, look at `app_probe/hello-readiness.yaml`.